src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/BuildUpToDateCheck.cs

```diff
@@ -0,0 +1,227 @@
+                {
+                    Log($"Disabled because the '{DisableFastUpToDateCheckProperty}' property is set to a non-empty value.");
+                    return false;
+                }
+
+                var inputs = new List<string>();
+
+                // add the project file
+                if (!string.IsNullOrEmpty(project.FullPath))
+                {
+                    inputs.Add(project.FullPath);
+                    Log($"Adding input project file {project.FullPath}.");
+                }
+
+                // add all imports except .user file
+                var userFilePath = project.FullPath + ".user";
+                var imports = project.Imports
+                    .Select(i => i.ImportedProject.FullPath)
+                    .Where(p => !string.IsNullOrEmpty(p) && !string.Equals(p, userFilePath, StringComparison.OrdinalIgnoreCase));
+
+                foreach (var import in imports)
+                {
+                    Log($"Adding input project import {import}.");
+                }
+
+                inputs.AddRange(imports);
+            
+                // add all project items (generally items seen in solution explorer) that are not excluded from UpToDate check
+                // Skip items that are marked as excluded from build.
+                var projectItemSchemaValue = (await _projectItemsSchema.GetSchemaAsync(cancellationToken)).Value;
+                var itemTypes = projectItemSchemaValue
+                    .GetKnownItemTypes()
+                    .Select(name => projectItemSchemaValue.GetItemType(name))
+                    .Where(item => item != null && item.UpToDateCheckInput && !string.Equals(item.Name, "None", StringComparison.OrdinalIgnoreCase));
+
+                foreach (var itemType in itemTypes)
+                {
+                    Log($"Checking known item type '{itemType.Name}'.");
+
+                    var items = project.GetItems(itemType.Name)
+                            .Where(item => !string.Equals(item.GetMetadataValue("ExcludedFromBuild"), "true", StringComparison.OrdinalIgnoreCase))
+                            .Select(item => item.GetMetadataValue("FullPath"));
+
+                    foreach (var item in items)
+                    {
+                        Log($"Input item path '{item}'.");
+                        inputs.Add(item);
+                    }
+                }
+
+                async Task CheckReferences<TUnresolvedReference, TResolvedReference>(string name, IResolvableReferencesService<TUnresolvedReference, TResolvedReference> service)
+                    where TUnresolvedReference : IProjectItem, TResolvedReference
+                    where TResolvedReference : class, IReference
+                {
+                    Log($"Checking reference type '{name}'.");
+                    if (service != null)
+                    {
+                        foreach (var resolvedReference in await service.GetResolvedReferencesAsync())
```

Comment:
Do we really get performance gain here?  if the result is not cached, each GetResolvedReferencesAsync can start a design time build, so it is possible that you will do extra 6 builds to save 1 build?